server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala

```diff
@@ -96,142 +100,306 @@ class AddressIndexRepository @Inject()(
-            matchQuery(
-              field = "lpi.townName",
-              value = token
-            )),
-          tokens.get(Tokens.postcode).map(token =>
-            matchQuery(
-              field = "lpi.postcodeLocator",
-              value = token
-            )),
-          tokens.get(Tokens.buildingNumber).map(token =>
-            matchQuery(
-              field = "paf.buildingNumber",
-              value = token
-            ).boost(queryParams.buildingNumberPafBoost)),
-          tokens.get(Tokens.organisationName).map(token =>
-            matchQuery(
-              field = "paf.organizationName",
-              value = token
-            )),
-          tokens.get(Tokens.departmentName).map(token =>
-            matchQuery(
-              field = "paf.departmentName",
-              value = token
-            )),
-          tokens.get(Tokens.subBuildingName).map(token =>
-            matchQuery(
-              field = "paf.subBuildingName",
-              value = token
-            ).boost(queryParams.subBuildingNameSubBuildingPafBoost)),
-          tokens.get(Tokens.subBuildingName).map(token =>
-            matchQuery(
-              field = "paf.buildingName",
-              value = token
-            ).boost(queryParams.subBuildingNameBuildingPafBoost)),
-          tokens.get(Tokens.buildingName).map(token =>
-            matchQuery(
-              field = "paf.buildingName",
-              value = token
-            )),
-          tokens.get(Tokens.streetName).map(token =>
-            fuzzyQuery(
-              name = "paf.thoroughfare",
-              value = token
-            ).boost(queryParams.streetNamePafBoost).fuzziness(Fuzziness.TWO)),
-          tokens.get(Tokens.townName).map(token =>
-            matchQuery(
-              field = "paf.postTown",
-              value = token
-            )),
-          tokens.get(Tokens.postcode).map(token =>
-            matchQuery(
-              field = "paf.postcode",
-              value = token
-            )),
-          tokens.get(Tokens.locality).map(token =>
-            matchQuery(
-              field = "paf.dependentLocality",
-              value = token
-            )),
-          Some(
-            matchQuery(
-              field = "_all",
-              value = tokens.map(_.value).mkString(" ")
-            ).boost(queryParams.underlineAllBoost)
-          )
-        ).flatten)
-      }.minimumShouldMatch(queryParams.minimumShouldMatch)
+      bool(
+        must(Seq(
+          bool(must(saoQuery).should(subBuildingNameQuery)),
+          bool(must(paoQuery).should(buildingNameQuery)),
+          bool(should(buildingNumberQuery)),
+          bool(should(streetNameQuery)),
+          bool(should(townNameQuery)),
+          bool(must(postcodeQuery).should(postcodeInOutQuery))
+        )).should(Seq(
+          bool(should(organisationNameQuery)),
+          bool(should(departmentNameQuery)),
+          bool(should(localityQuery)),
```

Comment:
Looks really good! In the future we may wish to consider changing the localityQuery to be more like townNameQuery because of the lack of definition for towns and localities. However, for time being, we should use this solution.